modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java

```diff
@@ -1087,14 +1085,20 @@ private void sendSessionAttributes(Map<?, ?> attrs, GridTaskSessionImpl ses)
 
         IgniteCheckedException ex = null;
 
+        UUID locNodeId = ctx.localNodeId();
+
         // Every job gets an individual message to keep track of ghost requests.
         for (ComputeJobSibling s : ses.getJobSiblings()) {
             GridJobSiblingImpl sib = (GridJobSiblingImpl)s;
 
             UUID nodeId = sib.nodeId();
 
-            // Pair can be null if job is finished.
-            if (rcvrs.remove(nodeId)) {
+            if (locNodeId.equals(nodeId)) {
+                // Local job notification.
+                ctx.job().onChangeTaskAttributes(ses.getId(), s.getJobId(), attrs);
+            }
+            else if (rcvrs.remove(nodeId)) {
```

Comment:
... but don't remove it after. Am I missing something?

Reply:
You're right, it's bad.